### PR TITLE
refactor: explore minimal dropzone

### PR DESCRIPTION
**Pull request by juliusmarminge**
Explores a minimal useDropzone as well as pulling it out to a separate package with utilities that can be reused for solid and vue implementations

This fits both the React version as well as **adding** a Solid implementation whilst still removing code 🤯 
CleanShot 2023-12-18 at 09 11 44@2x

Ref: UPL-66

### CODE REVIEW

File: packages/dropzone/src/react.tsx [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+ *   return (
+ *     <div {...getRootProps()}>
+ *       <input {...getInputProps()} />
+ *       <p>Drag and drop some files here, or click to select files</p>
+ *     </div>
+ *   )
+ * }
+ * ```
+ */
+export function useDropzone({
```
Comment (juliusmarminge) [inline packages/dropzone/src/react.tsx:56]: this should be easily portable to solid, vue etc now and can use all the helpers from `core.ts`

File: packages/dropzone/src/solid.tsx
Comment (juliusmarminge) [inline packages/dropzone/src/solid.tsx:1]: this should probably live in community package but proves the point that it's easy to make one that's not react and re-use a lot of the same functionality

File: packages/dropzone/src/core.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+export function isEnterOrSpace(event: { key?: string; keyCode?: number }) {
+  return (
+    ("key" in event && (event.key === " " || event.key === "Enter")) ||
+    ("keyCode" in event && (event.keyCode === 32 || event.keyCode === 13))
+  );
+}
+
+const isDefined = <T>(v: T | null | undefined): v is T => v != null;
+export function isValidSize(file: File, minSize: number, maxSize: number) {
+  if (isDefined(file.size)) {
```
Comment (t3dotgg) [inline packages/dropzone/src/core.ts:25]: Can we early escape if not defined maybe? Would clean up the nesting of this logic a bit

Comment (t3dotgg) [reply packages/dropzone/src/core.ts:25]: Oh is this code ripped from the lib? If so might be worth commenting on top

File: packages/dropzone/src/solid.tsx
Comment (t3dotgg) [reply packages/dropzone/src/solid.tsx:1]: You can keep it here for now and I'll deal with it

**Review by t3dotgg — APPROVED** [review]
Okay fiiiiiine

File: packages/dropzone/src/core.ts (hunk 2 above)
Comment (juliusmarminge) [reply packages/dropzone/src/core.ts:25]: Most of it is the same as before (i.e. forked from react-dropzone) and some of it I cleaned up a bit when removing stuff we didnt need

File: packages/dropzone/src/core.ts [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+export function isEnterOrSpace(event: { key?: string; keyCode?: number }) {
+  return (
+    ("key" in event && (event.key === " " || event.key === "Enter")) ||
+    ("keyCode" in event && (event.keyCode === 32 || event.keyCode === 13))
+  );
+}
+
+const isDefined = <T>(v: T | null | undefined): v is T => v != null;
+export function isValidSize(file: File, minSize: number, maxSize: number) {
```
Comment (juliusmarminge) [inline packages/dropzone/src/core.ts:31]: Fixed it @t3dotgg - not sure how much of this is even needed, files should always have a size on them and minSize and maxSize are typed as numbers, so there shouldn't be a case where those are nullish either...

can probably be a one-liner

```ts
export const isValidSize = (file: File, minSz: number, maxSz: number) => file.size >= minSz && file.size <= maxSz;
```

File: packages/react/src/components/dropzone.tsx [hunk 4]
```diff
@@ -1,6 +1,8 @@
 import { useCallback, useEffect, useState } from "react";
 import { twMerge } from "tailwind-merge";
 
+import type { FileWithPath } from "@uploadthing/dropzone/react";
```
Comment (markflorkowski) [inline packages/react/src/components/dropzone.tsx:4]: It feels a bit strange that this is where that type comes from, but probably fine.

**Review by markflorkowski — APPROVED** [review]

File: packages/react/src/components/dropzone.tsx (hunk 4 above)
Comment (juliusmarminge) [reply packages/react/src/components/dropzone.tsx:4]: i don't see where else we'd get it from, unless we just make it a generic File instead?

Comment (markflorkowski) [reply packages/react/src/components/dropzone.tsx:4]: We could have an uploadthing/types package, but that makes it weird elsewhere, and also makes maintenance a pain. It's fine as is.

Comment (juliusmarminge) [reply packages/react/src/components/dropzone.tsx]: I did standard File: https://github.com/pingdotgg/uploadthing/pull/524/commits/605fdb64819a74c51fdd07f513dbb4d96cfd7cdc

**Pull request merged**